lesson/outline/learning-standards.md

```diff
@@ -0,0 +1,411 @@
+
+Example: 
+
+```js
+function makeRequest(isSuccessful) {
+  return new Promise((resolve, reject) => {
+    isSuccessful ? resolve("Hurray!") : reject ("Boo");
+  })
+}
+
+// Using async/await
+test('the Promise should resolve', async () => {
+  await expect(makeRequest(true)).resolves.toBe('Hurray!');
+});
+
+test('the Promise should be rejected', async () => {
+  await expect(makeRequest(false)).rejects.toBe('Boo');
+});
+
+// Returning the Promise
+test('the Promise should resolve', () => {
+  return expect(makeRequest(true)).resolves.toBe('Hurray!');
+});
+
+test('the Promise should be rejected', () => {
+  return expect(makeRequest(false)).rejects.toBe('Boo');
+});
+```
+
+
+Sources:
+* https://www.pluralsight.com/guides/test-asynchronous-code-jest
+* https://jestjs.io/docs/asynchronous
+
+<hr>
+
+## Mocking functions with `jest.fn()`
+
+### Learning Standard Text
+
+The Jest library provides the `jest.fn()` function for creating a "mock" function. 
+* An optional implementation function may be passed to `jest.fn()` to define the mock function's behavior and return value. 
+* The mock function's behavior may be further specified using various methods provided to the mock function such as `.mockReturnValueOnce()`.
+* If no implementation function is provided, the mock function simply returns `undefined`. 
+* The mock function's usage (how it was called, what it returned, etc...) may be validated using the the [`expect()` API](https://jestjs.io/docs/expect).
+
+### Tags
+JavaScript, Testing, Jest
+
+### Additional Notes / Resources Used
+
+This example demonstrates how a mocked function may be used within a test to confirm the behavior of a higher-order function `executeCallback()`.
+
+```js
+// The function to test. It should call the callback function 3 times with the provided arguments.
+function executeCallbackThrice(fn, ...args) {
+  fn(...args);
+  fn(...args);
+  fn(...args);
+}
+
+test('mocking a callback function', () => {
+  // Define the mock function with an implementation.
+  const myMock = jest.fn(value => 'Hello ' + value);
+
+  // Pass myMock to executeCallback()
+  executeCallback(myMock, 'World');
+
+  // Make sure the the mocked function was called properly.
+  expect(myMock).toHaveBeenCalledTimes(3);
+  expect(myMock).toHaveBeenCalledWith('World');
+  expect(myMock).toHaveReturnedWith('Hello World');
+
+  // Set the next call to myMock() to return 'foobar'
+  myMock.mockReturnValueOnce('foobar');
+  
+  // pass myMock to executeCallback()
+  executeCallback(myMock, 'Codecademy');
+
+  expect(myMock).toHaveBeenCalledTimes(6);
+
+  // Check each individual call's return value (indexing starts at 1)
+  expect(myMock).toHaveNthReturnedWith(4, 'foobar');
+  expect(myMock).toHaveNthReturnedWith(5, 'Hello Codecademy');
+  expect(myMock).toHaveNthReturnedWith(6, 'Hello Codecademy');
+
+  // Reset all call data
+  myMock.mockClear();
+  expect(myMock).not.toHaveBeenCalled();
```

Comment:
Are we going to cover `beforeEach`? Usually this normalization logic happens there.

Reply:
I wasn't planning to. I'll just remove this from the example.